### PR TITLE
Tonic Interceptor

### PR DESCRIPTION
**Pull request by blogle**
Adds optional config for an interceptor to enable modifying each outbound metrics request.

For example - my company has a centralized opentelemetry-collector that gives our SRE team flexibility to determine where they should pipe and store metrics that feed into grafana. This central collector sits behind google IAP. Each internal service needs to supply an oauth token in the grpc headers to publish metrics to this endpoint. This PR exposes the machinery for applications to access the request object and supply such metadata.

### CODE REVIEW

File: opentelemetry-otlp/src/exporter/tonic.rs
Comment (jtescher) [inline opentelemetry-otlp/src/exporter/tonic.rs:29]: Nit: The custom debug impl could be moved to `BoxInterceptor`.

File: opentelemetry-otlp/src/exporter/tonic.rs
Comment (jtescher) [inline opentelemetry-otlp/src/exporter/tonic.rs:37]: This bound should be on the interceptor impl rather than assumed via unsafe.

```suggestion
pub(crate) struct BoxInterceptor(Box<dyn tonic::service::Interceptor + Send>);
```

File: opentelemetry-otlp/src/exporter/tonic.rs
Comment (jtescher) [inline opentelemetry-otlp/src/exporter/tonic.rs:45]: As mentioned above, not needed if `Debug` is implemented for `BoxInterceptor`.

**Review by jtescher — CHANGES_REQUESTED** [review]
Thanks for getting this started, makes sense to have interceptor support, added some comments.

File: opentelemetry-otlp/src/metric.rs
Comment (jtescher) [inline opentelemetry-otlp/src/metric.rs:304]: Remove extra println

**Comment by blogle** [comment]
bump @jtescher

Is anything else you would like to see on this PR? Getting this auth setup working is becoming a higher priority for our team, so would be great to get this integrated.

**Comment by TommyCpp** [comment]
Sorry didn't get a chance to take a look over the weekend. I will take a look before cut 0.19 release

**Comment by blogle** [comment]
Is this the only item going into the 0.19 release? If so what needs to be done to move this forward? We would really like to start migrating our metrics, but this has been a blocking issue for us.

**Review by hdost — APPROVED** [review]
Looks like you've addressed @jtescher suggestions but yes you'll need to need to sign the CLA. 
Thank you for your contribution!

**Review by TommyCpp — APPROVED** [review]

**Comment by TommyCpp** [comment]
Probably need to address the lint issue also. But I can take a look if needed

**Comment by blogle** [comment]
CLA should now be cleared up. @TommyCpp as for the lints, I believe those `pub` warnings are pre-existing as I don't think I touched visibility of anything. That said, if this is blocking the PR, I can probably make those all `pub(crate)` since `TonicExporterBuilder` is the actual public interface there.

**Comment by TommyCpp** [comment]
Looks like it needs your approval cc @jtescher

**Comment by TommyCpp** [comment]
I have pushed a change to fix the lint issue and make the configurations public

**Review by jtescher — APPROVED** [review]

**Pull request merged**
